File: examples/create.lua
local pandoc = require "pandoc"

local function Pandoc(doc)
  local function block(b)
    table.insert(doc.blocks, b)
  end

  local function inline(i)
    table.insert(doc.blocks, pandoc.Para {i})
  end

  block(pandoc.BlockQuote {
    pandoc.Para {
      pandoc.Str "a"
    }
  })

  block(pandoc.BulletList {
    {
      pandoc.Plain {
        pandoc.Str "a"
      }
    },
    {
      pandoc.Plain {
        pandoc.Str "b"
      }
    }
  })

  block(pandoc.CodeBlock "a")

  block(pandoc.DefinitionList {
    {
      {
        pandoc.Str "a"
      },
      {
        {
          pandoc.Plain {
            pandoc.Str "b"
          }
        }
      }
    },
    {
      {
        pandoc.Str "c"
      },
      {
        {
          pandoc.Plain {
            pandoc.Str "d"
          }
        }
      }
    }
  })

  block(pandoc.Div {
    pandoc.Para {
      pandoc.Str "a"
    }
  })

  block(pandoc.Header(2, {
    pandoc.Str "a"
  }))

  block(pandoc.HorizontalRule())
  block(pandoc.Null())

  block(pandoc.OrderedList {
    {
      pandoc.Plain {
        pandoc.Str "a"
      }
    },
    {
      pandoc.Plain {
        pandoc.Str "b"
      }
    }
  })

  block(pandoc.Para {
    pandoc.Str "a"
  })

  block(pandoc.Plain {
    pandoc.Str "a"
  })

  block(pandoc.RawBlock("html", "<p>a</p>"))

  inline(pandoc.Code "a")
  inline(pandoc.Emph { pandoc.Str "a" })

  inline(pandoc.Image(
    { pandoc.Str "a" },
    "b",
    "c"
  ))

  inline(pandoc.LineBreak())

  inline(pandoc.Link(
    { pandoc.Str "a" },
    "b",
    "c"
  ))

  inline(pandoc.Math("InlineMath", "a"))
  inline(pandoc.Math("DisplayMath", "a"))

  inline(pandoc.Note {
    pandoc.Para { pandoc.Str "a" }
  })

  inline(pandoc.Quoted(
    "SingleQuote",
    { pandoc.Str "a" }
  ))
  inline(pandoc.Quoted(
    "DoubleQuote",
    { pandoc.Str "a" }
  ))

  inline(pandoc.RawInline("html", "<p>a</p>"))
  inline(pandoc.SmallCaps { pandoc.Str "a" })
  inline(pandoc.SoftBreak())
  inline(pandoc.Space())
  inline(pandoc.Span { pandoc.Str "a" })
  inline(pandoc.Str "a")
  inline(pandoc.Strikeout { pandoc.Str "a" })
  inline(pandoc.Strong { pandoc.Str "a" })
  inline(pandoc.Superscript { pandoc.Str "a" })
  inline(pandoc.Subscript { pandoc.Str "a" })
  inline(pandoc.Underline { pandoc.Str "a" })
  return doc
end

return {
  {
    Pandoc = Pandoc
  }
}

File: examples/create.ts
#!/usr/bin/env node

import {
  create,
  filter,
  types
} from '../src/index.js'

const fs: filter.FilterSet = {
  Pandoc: function (doc) {
    const block = (b: types.Block) => doc.blocks.push(b)
    const inline = (i: types.Inline) => doc.blocks.push(create.Para([i]))

    // Test blocks
    block(
      create.BlockQuote([
        create.Para([
          create.Str('a')
        ])
      ])
    )

    block(
      create.BulletList([
        [
          create.Plain([
            create.Str('a')
          ])
        ],
        [
          create.Plain([
            create.Str('b')
          ])
        ]
      ])
    )

    block(create.CodeBlock('a'))

    block(
      create.DefinitionList([
        [
          [
            create.Str('a')
          ],
          [
            [
              create.Plain([
                create.Str('b')
              ])
            ]
          ]
        ],
        [
          [
            create.Str('c')
          ],
          [
            [
              create.Plain([
                create.Str('d')
              ])
            ]
          ]
        ]
      ])
    )

    block(
      create.Div([
        create.Para([
          create.Str('a')
        ])
      ])
    )

    block(
      create.Header(2, [
        create.Str('a')
      ])
    )

    block(create.HorizontalRule())

    // TODO LineBlock

    block(create.Null())

    block(
      create.OrderedList([
        [
          create.Plain([
            create.Str('a')
          ])
        ],
        [
          create.Plain([
            create.Str('b')
          ])
        ]
      ])
    )

    block(
      create.Para([
        create.Str('a')
      ])
    )

    block(
      create.Plain([
        create.Str('a')
      ])
    )

    block(create.RawBlock('html', '<p>a</p>'))

    // TODO Table

    // Test inlines

    // TODO Cite

    inline(create.Code('a'))

    inline(
      create.Emph([
        create.Str('a')
      ])
    )

    inline(
      create.Image(
        [
          create.Str('a')
        ],
        'b',
        'c'
      )
    )

    inline(create.LineBreak())

    inline(
      create.Link(
        [
          create.Str('a')
        ],
        'b',
        'c'
      )
    )

    inline(create.Math(types.MathType.InlineMath, 'a'))
    inline(create.Math(types.MathType.DisplayMath, 'a'))

    inline(
      create.Note([
        create.Para([
          create.Str('a')
        ])
      ])
    )

    inline(
      create.Quoted(
        types.QuoteType.SingleQuote,
        [
          create.Str('a')
        ]
      )
    )
    inline(
      create.Quoted(
        types.QuoteType.DoubleQuote,
        [
          create.Str('a')
        ]
      )
    )

    inline(create.RawInline('html', '<p>a</p>'))

    inline(
      create.SmallCaps([
        create.Str('a')
      ])
    )

    inline(create.SoftBreak())
    inline(create.Space())

    inline(
      create.Span([
        create.Str('a')
      ])
    )

    inline(create.Str('a'))

    inline(
      create.Strikeout([
        create.Str('a')
      ])
    )
    inline(
      create.Strong([
        create.Str('a')
      ])
    )
    inline(
      create.Superscript([
        create.Str('a')
      ])
    )
    inline(
      create.Subscript([
        create.Str('a')
      ])
    )
    inline(
      create.Underline([
        create.Str('a')
      ])
    )
    return doc
  }
}

filter.interact([fs])
